core/src/main/scala/com/pingcap/tispark/TiDBRelation.scala

```diff
@@ -65,13 +65,13 @@ case class TiDBRelation(
 
   def logicalPlanToRDD(dagRequest: TiDAGRequest, output: Seq[Attribute]): List[TiRowRDD] = {
     import scala.collection.JavaConverters._
-    val ids = dagRequest.getIds.asScala
+    val ids = dagRequest.getPrunedPhysicalIds.asScala
     var tiRDDs = new ListBuffer[TiRowRDD]
     val tiConf = session.getConf
     tiConf.setPartitionPerSplit(TiUtil.getPartitionPerSplit(sqlContext))
     ids.foreach(id => {
       tiRDDs += new TiRowRDD(
-        dagRequest,
+        dagRequest.copyReqWithPhysicalId(id),
```

Comment:
Request for each partition should only contain key ranges of the specific physical id.

The previous approach will contain all key ranges, that will be a N*N ranges in all.